src/libopensc/compression.c

```diff
@@ -192,21 +195,21 @@ static int sc_decompress_zlib_alloc(u8** out, size_t* outLen, const u8* in, size
 		gz.avail_out = bufferSize - *outLen;
 
 		err = inflate(&gz, Z_FULL_FLUSH);
-		if(err != Z_STREAM_END && err != Z_OK) {
-			if(*out)
-				free(*out);
+		if (err != Z_STREAM_END && err != Z_OK) {
+			free(*out);
 			*out = NULL;
 			break;
 		}
-		num = bufferSize - *outLen - gz.avail_out;
-		if(num > 0) {
-			*outLen += num;
-			bufferSize += num + blockSize;
+		num = *outLen + gz.avail_out;
+		if (bufferSize > num) {
+			*outLen += bufferSize - num;
+			bufferSize += bufferSize - num + blockSize;
 		}
```

Comment:
Indeed, that's what I deduced as well. No matter if we look at the bytes consumed or the bytes left, we rely on zlib to set these values correctly. So from our point of view both operations are equivalent. I agree, that your suggestion would be more readable, but since I don't have something for testing, I didn't want to rewrite the code beyond avoiding the integer underflow.

Is this covered in the test suite?

Reply:
Right. After another read, it looks like it makes sense. The `outLen` is the currently occupied data in `buf` and adding `gz.avail_out` checks whether there is enough space in the allocated `bufferSize`. This is right, but it would deserve comment.

Math is though confusing after that. I think the increment would be equivalent with following (maybe more readable):
```
			*outLen = bufferSize - gz.avail_out;
```
The increment of `bufferSize` is (I think) arbitrary input to the next realloc so I think it does not need to be that complicated either. Something like this should be sufficient:
```
			bufferSize += blockSize;
```

The testsuite invokes `sc_decompress_alloc()`, which invokes this function. Feel free to extend it of appropriate test cases. It would be nice to add few more test vectors that would execute these code paths (sufficiently long valid data blocks).

Reply:
We could switch to ag's implementation, which allocates to page size boundary...
https://github.com/ggreer/the_silver_searcher/blob/9163065a5efcedc2da480ff0547bbe6bc41e5221/src/decompress.c#L27-L100